File: design-pattern-go/template_method/template_method.go
package template_method

import "fmt"

type Printer interface {
	open()
	print()
	close()
}

/*****************************
 * AbstractDisplay
 ****************************/
type AbstractDisplay struct {
	printer Printer
}

func (a *AbstractDisplay) Display() {
	a.printer.open()
	for i := 0; i < 5; i++ {
		a.printer.print()
	}
	a.printer.close()
}

/*****************************
 * CharDisplay
 ****************************/
type CharDisplay struct {
	*AbstractDisplay
	ch rune
}

func NewCharDispaly(ch rune) *CharDisplay {
	charDisplay := &CharDisplay{
		AbstractDisplay: &AbstractDisplay{},
		ch:              ch,
	}
	charDisplay.printer = charDisplay
	return charDisplay
}

func (c *CharDisplay) open() {
	fmt.Print("<<")
}

func (c *CharDisplay) print() {
	fmt.Print(string(c.ch))
}

func (c *CharDisplay) close() {
	fmt.Print(">>")
}

/*****************************
 * StringDisplay
 ****************************/
type StringDisplay struct {
	*AbstractDisplay
	str   string
	width int
}

func NewStringDispaly(str string) *StringDisplay {
	stringDisplay := &StringDisplay{
		AbstractDisplay: &AbstractDisplay{},
		str:             str,
		width:           len(str),
	}
	stringDisplay.printer = stringDisplay
	return stringDisplay
}

func (s *StringDisplay) open() {
	s.printLine()
}

func (s *StringDisplay) print() {
	fmt.Printf("%s%s%s\n", "|", s.str, "|")
}

func (s *StringDisplay) close() {
	s.printLine()
}

func (s *StringDisplay) printLine() {
	fmt.Print("+")
	for i := 0; i < s.width; i++ {
		fmt.Print("-")
	}
	fmt.Println("+")
}

File: design-pattern-python/design_pattern_python/template_method/template_method.py
from abc import ABCMeta, abstractmethod


class AbstractDisplay(metaclass=ABCMeta):
    @abstractmethod
    def begin(self) -> None:
        pass

    @abstractmethod
    def output(self) -> None:
        pass

    @abstractmethod
    def end(self) -> None:
        pass

    def display(self) -> None:
        self.begin()
        for i in range(5):
            self.output()
        self.end()


class CharDisplay(AbstractDisplay):
    def __init__(self, ch: str) -> None:
        self.__ch = ch

    def begin(self) -> None:
        print("<<", end="")

    def output(self) -> None:
        print(self.__ch, end="")

    def end(self) -> None:
        print(">>")


class StringDisplay(AbstractDisplay):
    def __init__(self, string: str) -> None:
        self.__string = string
        self.__width = len(string)

    def begin(self) -> None:
        self.print_line()

    def output(self) -> None:
        print("|" + self.__string + "|")

    def end(self) -> None:
        self.print_line()

    def print_line(self) -> None:
        print("+", end="")
        for _ in range(self.__width):
            print("-", end="")
        print("+")


if __name__ == "__main__":
    d1 = CharDisplay("H")
    d2 = StringDisplay("Hello, world.")

    d1.display()
    print("")
    d2.display()
